frame/multisig/src/lib.rs

```diff
@@ -524,6 +519,163 @@ impl<T: Trait> Module<T> {
+	) -> DispatchResultWithPostInfo {
+		ensure!(threshold >= 2, Error::<T>::ZeroThreshold);
+		let max_sigs = T::MaxSignatories::get() as usize;
+		ensure!(!other_signatories.is_empty(), Error::<T>::TooFewSignatories);
+		let other_signatories_len = other_signatories.len();
+		ensure!(other_signatories_len < max_sigs, Error::<T>::TooManySignatories);
+		let signatories = Self::ensure_sorted_and_insert(other_signatories, who.clone())?;
+
+		let id = Self::multi_account_id(&signatories, threshold);
+
+		// Threshold > 1; this means it's a multi-step operation. We extract the `call_hash`.
+		let (call_hash, call_len, maybe_call, store) = match call_or_hash {
+			CallOrHash::Call(call, should_store) => {
+				let call_hash = blake2_256(&call);
+				let call_len = call.len();
+				(call_hash, call_len, Some(call), should_store)
+			}
+			CallOrHash::Hash(h) => (h, 0, None, false),
+		};
+
+		// Branch on whether the operation has already started or not.
+		if let Some(mut m) = <Multisigs<T>>::get(&id, call_hash) {
+			// Yes; ensure that the timepoint exists and agrees.
+			let timepoint = maybe_timepoint.ok_or(Error::<T>::NoTimepoint)?;
+			ensure!(m.when == timepoint, Error::<T>::WrongTimepoint);
+
+			// Ensure that either we have not yet signed or that it is at threshold.
+			let mut approvals = m.approvals.len() as u16;
+			// We only bother with the approval if we're below threshold.
+			let maybe_pos = m.approvals.binary_search(&who).err().filter(|_| approvals < threshold);
+			// Bump approvals if not yet voted and the vote is needed.
+			if maybe_pos.is_some() { approvals += 1; }
+
+			// We only bother fetching/decoding call if we know that we're ready to execute.
+			let maybe_approved_call = if approvals >= threshold {
+				Self::get_call(&call_hash, maybe_call.as_ref().map(|c| c.as_ref()))
+			} else { None };
+
+			if let Some(call) = maybe_approved_call {
+				// verify weight
+				ensure!(call.get_dispatch_info().weight <= max_weight, Error::<T>::WeightTooLow);
+				// We're now executing as a freshly authenticated new account, so the previous call
+				// restrictions no longer apply.
+				let _guard = ClearFilterGuard::<T::IsCallable, <T as Trait>::Call>::new();
+				ensure!(T::IsCallable::filter(&call), Error::<T>::Uncallable);
+
+				let result = call.dispatch(frame_system::RawOrigin::Signed(id.clone()).into());
+				let _ = T::Currency::unreserve(&m.depositor, m.deposit);
+				<Multisigs<T>>::remove(&id, call_hash);
+				Self::clear_call(&call_hash);
+				Self::deposit_event(RawEvent::MultisigExecuted(
+					who, timepoint, id, call_hash, result.map(|_| ()).map_err(|e| e.error)
+				));
+				Ok(None.into())
+			} else {
+				// We cannot dispatch the call now; either it isn't available, or it is, but we
+				// don't have threshold approvals even with our signature.
+
+				// Store the call if desired.
+				let stored = maybe_call
+					.filter(|_| store)
+					.map_or(false, |data| Self::store_call(who.clone(), &call_hash, data));
```

Comment:
I don't think we have written to storage, yet, so it should be fine to error out of `store_call` if there are not enough funds.

Reply:
Here, yes, but below it's impossible, and I'd prefer consistency.